File: #0525.contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {

        int size = nums.size();
        unordered_map<int, int> m;
        m[0] = -1;
        int result = 0;
        int index = 0;
        for (int i = 0; i < size; i++) {
            index += (nums[i] == 0 ? -1 : 1);
            if (m.find(index) != m.end()) {
                result = std::max(result, i - m[index]);
            } else {
                m[index] = i;
            }
        }
        return result;
    }
};

File: #0525.contiguous-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMaxLength = function(nums) {
    let count = 0;
    let max = 0;
    let hash = { 0:-1 };
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] === 0) {
            count--;
        } else {
            count++;
        }

        if (hash[count] != null) {
            max = Math.max(max, i - hash[count]);
        } else {
            hash[count] = i;
        }
    }  
    return max;
};
